File: web/app/data-health.js
import { $, readJSONOrText } from "./shared.js";

let busy = false;
let lastReport = null;
let lastReceivedAt = null;
let lastAttemptAt = 0;
let deliveryFailed = false;

export function updateDataHealthReceipt() {
  const receipt = $("dataHealthReceipt");
  if (!receipt || busy) return;
  if (deliveryFailed || (lastReport && !(Date.parse(lastReport.valid_until) > Date.now()))) {
    receipt.textContent = lastReport ? "Showing the last Canary assessment. A current report has not been received; current health is unconfirmed." : "Canary's report could not be received. Source health is unverified.";
  } else if (lastReport && lastReceivedAt) {
    receipt.textContent = `Report received ${lastReceivedAt.toLocaleTimeString()}. Canary checked it ${new Date(lastReport.as_of).toLocaleTimeString()}.`;
  }
}

// Rows and counts are daemon-authored. Only receipt/error state belongs here.
export async function refreshDataHealth({ minIntervalMs = 0 } = {}) {
  if (busy || Date.now() - lastAttemptAt < minIntervalMs) return;
  lastAttemptAt = Date.now();
  busy = true;
  const receipt = $("dataHealthReceipt");
  if (!receipt) { busy = false; return; }
  receipt.textContent = "Reading Canary's report…";
  try {
    let first = null;
    let offset = 0;
    const sources = [];
    for (let page = 0; page < 128; page++) {
      const query = new URLSearchParams({ offset: String(offset), limit: "24" });
      if (first) query.set("revision", first.revision);
      const response = await fetch(`/api/data/health?${query}`, { credentials: "include", signal: AbortSignal.timeout(10000) });
      const report = await readJSONOrText(response);
      if (!response.ok || report?.schema_version !== 1 || !report.summary || !Array.isArray(report.sources) || report.offset !== offset || (first && report.revision !== first.revision)) throw new Error("Canary report unavailable or unsupported");
      if (!first) first = report;
      sources.push(...report.sources);
      if (report.complete) {
        if (sources.length !== first.summary.total) throw new Error("Incomplete Canary source coverage");
        lastReport = { ...first, sources };
        lastReceivedAt = new Date();
        deliveryFailed = false;
        renderDataHealth(lastReport);
        return;
      }
      if (!Number.isInteger(report.next_offset) || report.next_offset !== sources.length || report.next_offset <= offset) throw new Error("Incomplete Canary report");
      offset = report.next_offset;
    }
    throw new Error("Report coverage exceeds the display limit");
  } catch {
    deliveryFailed = true;
  } finally { busy = false; updateDataHealthReceipt(); }
}

export function renderDataHealth(report) {
  const summary = $("dataHealthSummary");
  const list = $("dataHealthSources");
  if (!summary || !list) return;
  summary.textContent = report.summary.label;
  list.replaceChildren();
  for (const source of report.sources) {
    const item = document.createElement("details");
    item.className = "data-health-source";
    item.dataset.state = source.state || "unknown";
    const title = document.createElement("summary");
    title.textContent = `${source.name} — ${source.receiving || "Unverified"}`;
    item.append(title);
    const text = [source.provider, ...(source.affects || []), source.detail, source.action];
    if (source.access) text.push(`Live access: ${source.access.reason} (IBKR ${source.access.code}). Retry ${new Date(source.access.retry_at).toLocaleTimeString()}.`);
    if (source.source_at) text.push(`Source time: ${new Date(source.source_at).toLocaleString()}`);
    else if (source.source_date) text.push(`Source date: ${source.source_date}`);
    else text.push("Source time: unknown");
    if (source.received_at) text.push(`Data received: ${new Date(source.received_at).toLocaleString()}`);
    if (source.checked_at) text.push(`Checked: ${new Date(source.checked_at).toLocaleString()}`);
    for (const line of text.filter(Boolean)) {
      const p = document.createElement("p");
      p.textContent = line;
      item.append(p);
    }
    list.append(item);
  }
}

File: internal/rpc/data_health.go
package rpc

import "time"

// Data-health methods serve producer-owned observations; only Check requests work.
const (
	MethodDataHealth        = "data.health"
	MethodDataCheck         = "data.check"
	DataHealthSchemaVersion = 1
)

// DataHealthParams selects a bounded page of one report revision.
type DataHealthParams struct {
	Offset   int    `json:"offset,omitempty"`
	Limit    int    `json:"limit,omitempty"`
	Revision string `json:"revision,omitempty"`
}

// DataHealthResult is the daemon's authoritative data-health report. Adapters
// preserve its assessment; receipt or browser health is a separate observation.
type DataHealthResult struct {
	SchemaVersion int                 `json:"schema_version"`
	Revision      string              `json:"revision"`
	AsOf          time.Time           `json:"as_of"`
	ValidUntil    time.Time           `json:"valid_until"`
	ScopeState    string              `json:"scope_state"`
	Check         *DataCheckResult    `json:"check,omitempty"`
	Summary       DataHealthSummary   `json:"summary"`
	Concerns      []DataHealthConcern `json:"concerns"`
	Sources       []DataSourceHealth  `json:"sources"`
	Offset        int                 `json:"offset"`
	NextOffset    *int                `json:"next_offset,omitempty"`
	Complete      bool                `json:"complete"`
}

// DataHealthSummary counts producer services/data products and distinct observed causes.
// Instruments and requested chart windows never create sources or source problems.
// Unverified is separate from known failures; a complete page is not coverage.
type DataHealthSummary struct {
	State       string `json:"state"`
	Label       string `json:"label"`
	Total       int    `json:"total"`
	Required    int    `json:"required"`
	Current     int    `json:"current"`
	Limited     int    `json:"limited"`
	Unavailable int    `json:"unavailable"`
	Unverified  int    `json:"unverified"`
	Problems    int    `json:"problems"`
}

// DataSourceHealth separates data actually received from the latest access
// attempt and its impact. IDs identify services or data products, never individual
// instruments. Instrument availability and chart coverage stay in their own DTOs.
// It contains no symbols, prices, balances or broker free text.
type DataSourceHealth struct {
	ID       string `json:"id"`
	Name     string `json:"name"`
	Provider string `json:"provider"`
	Kind     string `json:"kind"`
	Required bool   `json:"required"`
	State    string `json:"state"`
	// Availability reports observed acquisition, independently of delivery mode,
	// scheduled cadence, applicability and analytical usability. Absent optional
	// dimensions mean unsupported; they must never be interpreted as healthy.
	Availability     string                 `json:"availability,omitempty"`
	CadenceState     string                 `json:"cadence_state,omitempty"`
	Applicability    string                 `json:"applicability,omitempty"`
	Usability        string                 `json:"usability,omitempty"`
	UsabilityReason  string                 `json:"usability_reason,omitempty"`
	Delivery         string                 `json:"delivery,omitempty"`
	Receiving        string                 `json:"receiving"`
	Detail           string                 `json:"detail,omitempty"`
	Affects          []string               `json:"affects"`
	DataType         string                 `json:"data_type,omitempty"`
	SourceAt         time.Time              `json:"source_at,omitzero"`
	SourceDate       string                 `json:"source_date,omitempty"`
	SourceTimeKind   string                 `json:"source_time_kind,omitempty"`
	ReceivedAt       time.Time              `json:"received_at,omitzero"`
	CheckedAt        time.Time              `json:"checked_at,omitzero"`
	ValidUntil       time.Time              `json:"valid_until,omitzero"`
	NextAttempt      time.Time              `json:"next_attempt,omitzero"`
	Action           string                 `json:"action,omitempty"`
	Access           *DataAccessObservation `json:"access,omitempty"`
	Failure          *SourceFailure         `json:"failure,omitempty"`
	ProblemIDs       []string               `json:"problem_ids,omitempty"`
	DerivedFrom      []string               `json:"derived_from,omitempty"`
	FirstObserved    time.Time              `json:"first_observed,omitzero"`
	LastSuccess      time.Time              `json:"last_success,omitzero"`
	History          []DataHealthTransition `json:"history,omitempty"`
	HistoryTruncated bool                   `json:"history_truncated,omitempty"`
}

// DataHealthTransition is historical diagnostic evidence, never restored access
// authority. Observation gaps and session changes cannot establish continuity.
type DataHealthTransition struct {
	At       time.Time `json:"at"`
	State    string    `json:"state"`
	DataType string    `json:"data_type,omitempty"`
	Reason   string    `json:"reason,omitempty"`
}

// DataCheckResult counts checked source services, not individual reference probes.
// It reports bounded acquisition outcomes, not subscription repair.
type DataCheckResult struct {
	StartedAt   time.Time `json:"started_at"`
	CompletedAt time.Time `json:"completed_at,omitzero"`
	Running     bool      `json:"running"`
	Reused      int       `json:"reused"`
	Checked     int       `json:"checked"`
	Limited     int       `json:"limited"`
	Failed      int       `json:"failed"`
	Deferred    int       `json:"deferred"`
	Detail      string    `json:"detail"`
}

// HealthVerdict is daemon-authored operational posture, separate from trade
// eligibility. Client-version or local delivery diagnostics stay outside it.
type HealthVerdict struct {
	State  string `json:"state"`
	Reason string `json:"reason"`
}

// DataAccessObservation is a historical access attempt, separate from received data.
type DataAccessObservation struct {
	Code       int       `json:"code"`
	Reason     string    `json:"reason"`
	ObservedAt time.Time `json:"observed_at"`
	RetryAt    time.Time `json:"retry_at"`
}

// DataHealthConcern is a bounded producer-ranked concern. Unknown coverage is
// distinct from a confirmed problem and never increments the problem count.
type DataHealthConcern struct {
	SourceID string `json:"source_id"`
	State    string `json:"state"`
	Label    string `json:"label"`
}
